File: languages/typescript/runtime/utils/brand.ts
import { readdirSync, statSync } from 'node:fs'
import { join } from 'node:path'

export function inferBrandRoot(root: string): string {
  const candidates = readdirSync(root)
    .map((name) => join(root, name))
    .filter((path) => {
      const name = path.split('/').at(-1) ?? ''
      if (
        (!name.startsWith('.') && name !== '__BRAND_ROOT__') ||
        name.endsWith('-plugin')
      ) {
        return false
      }
      if (!statSync(path).isDirectory()) {
        return false
      }
      return ['settings.json', 'settings.local.json', 'instructions.md', 'sessions'].some((child) =>
        exists(join(root, name, child)),
      )
    })
    .sort()

  return candidates[0] ?? join(root, '.claude')
}

export function instructionCandidates(root: string, brandRoot: string): string[] {
  const out: string[] = []
  for (const name of readdirSync(root).sort()) {
    if (!name.endsWith('.md') || ['README.md', 'AGENTS.md', 'CLAUDE.md'].includes(name)) {
      continue
    }
    out.push(join(root, name))
  }
  for (const name of ['AGENTS.md', 'CLAUDE.md']) {
    const path = join(root, name)
    if (exists(path)) out.push(path)
  }
  for (const name of ['instructions.md']) {
    const path = join(brandRoot, name)
    if (exists(path)) out.push(path)
  }
  for (const name of readdirSync(brandRoot).sort()) {
    if (!name.endsWith('.md') || name === 'instructions.md') {
      continue
    }
    out.push(join(brandRoot, name))
  }
  return out
}

function exists(path: string): boolean {
  try {
    return statSync(path) !== undefined
  } catch {
    return false
  }
}

File: languages/python/runtime/brand.py
from __future__ import annotations

from pathlib import Path


def infer_brand_root(project_root: Path) -> Path:
    candidates: list[Path] = []
    for path in project_root.iterdir():
        if not path.is_dir():
            continue
        if not path.name.startswith(".") or path.name.endswith("-plugin"):
            continue
        score = 0
        for child in ["settings.json", "settings.local.json", "instructions.md", "sessions"]:
            if (path / child).exists():
                score += 1
        if score:
            candidates.append(path)
    if candidates:
        candidates.sort(key=lambda path: path.name)
        return candidates[0]
    return project_root / ".claude"


def brand_doc_candidates(project_root: Path, brand_root: Path) -> list[Path]:
    candidates: list[Path] = []
    for name in ["AGENTS.md", "CLAUDE.md"]:
        path = project_root / name
        if path.exists():
            candidates.append(path)
    for path in project_root.glob("*.md"):
        if path.name in {"README.md", "AGENTS.md", "CLAUDE.md"}:
            continue
        candidates.insert(0, path)
    for name in ["instructions.md"]:
        path = brand_root / name
        if path.exists():
            candidates.append(path)
    for path in brand_root.glob("*.md"):
        if path.name == "instructions.md":
            continue
        candidates.append(path)
    return candidates
